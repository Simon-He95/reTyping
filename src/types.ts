export interface InsertPatch {
  type: 'insert'
  from: number
  text: string
}

export interface RemovalPatch {
  type: 'removal'
  from: number
  length: number
}

export type Patch = InsertPatch | RemovalPatch
