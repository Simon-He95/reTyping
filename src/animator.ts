import { Patch } from "."

function randomRgb() {
  return `rgb(${Math.floor(Math.random() * 255)}, ${Math.floor(
    Math.random() * 255
  )}, ${Math.floor(Math.random() * 255)})`
}

export function* createAnimator(input: string, patches: Patch[]) {
  let output = input
  let cursor = 0
  for (const patch of patches) {
    if (patch.type === 'insert') {
      cursor = patch.from
      const head = output.slice(0, patch.from)
      const tail = output.slice(patch.from)
      let selection = ''
      for (const chart of patch.text) {

        selection += `<span style="color:${randomRgb()}">${chart}</span>`
        yield { cursor: cursor + selection.length, output: head + selection + tail }
      }
      output = head + patch.text + tail
    }
    else if (patch.type === 'removal') {
      cursor = patch.from - patch.length
      const head = output.slice(0, patch.from - patch.length)
      const tail = output.slice(patch.from)
      let selection = output.slice(patch.from - patch.length, patch.from)
      for (let i = selection.length; i >= 0; i--) {
        yield { cursor: cursor + i, output: head + selection.slice(0, i) + tail }
      }
      output = head + tail

    }
  }
}