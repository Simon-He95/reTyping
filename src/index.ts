export * from './animator'
export * from './types'
export * from './patch'
