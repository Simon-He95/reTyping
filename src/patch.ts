import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'
import type { Patch } from './types'

export function diff(a: string, b: string): Diff[] {
  const differ = new DMP()
  const delta = differ.diff_main(a, b)
  differ.diff_cleanupSemantic(delta)
  return delta
}

export function calculatePatch(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  let index = 0
  for (const change of diffs) {
    if (change[0] === 0) {
      index += change[1].length
      continue
    }
    if (change[0] === -1) {
      patches.push({
        type: 'removal',
        from: index + change[1].length,
        length: change[1].length,
      })
    }
    if (change[0] === 1) {
      patches.push({
        type: 'insert',
        from: index,
        text: change[1],
      })
      index += change[1].length
    }
  }
  return patches
}

export function applyPatches(input: string, patches: Patch[]) {
  let text = input
  for (const patch of patches) {
    if (patch.type === 'insert')
      text = text.slice(0, patch.from) + patch.text + text.slice(patch.from)

    else if (patch.type === 'removal')
      text = text.slice(0, patch.from - patch.length) + text.slice(patch.from)
  }
  return text
}
