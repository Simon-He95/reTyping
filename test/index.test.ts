import { describe, expect, it } from 'vitest'
import { applyPatches, calculatePatch, diff } from '../src'
import { input, output } from './fixture'

describe('should', () => {
  it('exported', () => {
    const delta = diff(input, output)
    expect(delta).toMatchSnapshot('delta')
    const patches = calculatePatch(delta)
    expect(patches).toMatchSnapshot('patches')
    const applied = applyPatches(input, patches)
    expect(applied).toMatchSnapshot()
  })
})
