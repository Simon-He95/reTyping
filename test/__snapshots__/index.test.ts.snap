// Vitest Snapshot v1

exports[`should > exported > delta 1`] = `
[
  [
    0,
    "
import { describe, expect, it } from \\"vitest\\";
import { one",
  ],
  [
    1,
    ",two",
  ],
  [
    0,
    " } from '../src'
describe('should', () => {
  it('",
  ],
  [
    -1,
    "exported",
  ],
  [
    1,
    "one",
  ],
  [
    0,
    "', () => {
    expect(one).toEqual(1)
",
  ],
  [
    1,
    "    expect(two).toEqual(2)
",
  ],
  [
    0,
    "  })
})
",
  ],
]
`;

exports[`should > exported > patches 1`] = `
[
  {
    "from": 60,
    "text": ",two",
    "type": "insert",
  },
  {
    "from": 122,
    "length": 8,
    "type": "removal",
  },
  {
    "from": 114,
    "text": "one",
    "type": "insert",
  },
  {
    "from": 155,
    "text": "    expect(two).toEqual(2)
",
    "type": "insert",
  },
]
`;

exports[`should > exported 1`] = `
"
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
"
`;
